support non-droppable menus in extensions

cwillisf opened the pull request:
### Resolves

Resolves #709

### Proposed Changes

When an extension defines a menu, the two existing definition forms are still allowed:
```js
menuWithStaticItems: ['a', 'b'],
menuWithDynamicItems: 'getDynamicItems'
```

This pull request adds support for defining a menu with an object, and makes the above forms shorthand for an object which specifies only an `items` property. These definitions are equivalent to those above:
```js
menuWithStaticItems: { items: ['a', 'b'] },
menuWithDynamicItems: { items: 'getDynamicItems' }
```

Doing so means we can also add more settings to the object. This pull request adds support for a `rejectsReporters` flag which makes the menu "non-droppable" -- it won't accept reporters:
```js
nonDroppableMenu: { items: 'getVariableNames', rejectReporters: true }
```

**I am open to suggestions for a better name than `rejectReporters`.** I would prefer to keep it a "positive" flag such that omitting the flag or setting it to a falsey value both result in a droppable menu by default.

### Reason for Changes

There are a few menus in Scratch 3.0 which don't accept reporters, such as the variables and lists menus and the key selection menu for the "when key pressed" hat block. Extensionifying those blocks means supporting non-droppable menus.

CC @ericrosenbaum @evhan55

Commit:
support non-droppable menus in extensions

towerofnix commented on docs/extensions.md:
```diff
@@ -311,7 +311,16 @@ class SomeBlocks {
+                // The examples above are shorthand for setting only the `items` property in this full form:
+                menuC: {
+                    // This flag makes a "non-droppable" menu: the menu will not accept reporters.
+                    rejectReporters: true,
```
Frankly, save for someone hypothetically confusing the term with "reject" in Promise lingo, I think `rejectReporters` isn't too bad.. :P

Another option would be something along the lines of "only" - `onlyDropdown`, `onlyField`, etc. But I think `rejectReporters` is more clear than those. There's also `disallowReporters`, but that sort of screams, "It *should* be `allowReporters: false`!" - negation in the value rather than the name, that is (and `: false` is already explicitly out of the question).

PS a fun thing with making menus be in object form is that we can add entries like `changeCallback`, for implementing things like the "stop (all/this script/other scripts in sprite)" block!

Commit:
switch from `rejectReporters` flag to `acceptReporters`

Also add more extensive documentation around extension menu definitions,
including some considerations to think about before making a menu accept
reporters.

cwillisf commented:
After discussion in a meeting earlier today, we decided that defaulting to non-droppable menus would be best.

While ideally most or all drop-down menus would accept reporters, it's easy for that to get confusing or even frustrating unless the extension author thinks through a lot of details. To avoid bad experiences being the norm for dropdown menus in extensions, we decided to make this an opt-in feature.

I added more documentation detail on this topic in 88da71474a25adca9864ca1b10933e3a386ea5ec, including a few of the specific considerations to think through before making a menu accept reporters.

CC @thisandagain @ericrosenbaum @knandersen @kchadha @evhan55 @cwillisf

Commit:
fix tests: empty `menus` shouldn't be an array

Commit:
for compatibility, make all existing extension menus accept reporters

cwillisf commented:
Since 642a8c09b09b1997be16382d2ea623b95d1f3322 involves a lot of indentation changes I encourage reviewers to check this box:
![image](https://user-images.githubusercontent.com/7019101/56924373-98e54b00-6a81-11e9-8f4a-cadb1a6eee6a.png)

kchadha commented on docs/extensions.md:
```diff
@@ -120,6 +120,123 @@ class SomeBlocks {
+  * In particular, changing languages should never break a working project.
+* The average Scratch user should be able to figure out the valid values for this input without referring to extension
+  documentation.
+  * One way to ensure this is to make an item's text match or include the item's value.
```
Maybe it makes sense to include an example of this (e.g. the Music Extension lists the extension number as part of the menu item text)?

kchadha commented on src/engine/runtime.js:
```diff
@@ -830,18 +825,24 @@ class Runtime extends EventEmitter {
      */
     _fillExtensionCategory (categoryInfo, extensionInfo) {
+        categoryInfo.blocks = [];
+        categoryInfo.customFieldTypes = {};
```
Any particular reason for moving this out of `_registerExtensionPrimitives` and moving it here instead?

cwillisf replied:
`_fillExtensionCategory` is also called by `_refreshExtensionPrimitives` so we need to clear these out to prevent duplicating things on refresh. Previously `_refreshExtensionPrimitives` was clearing blocks and menus but not custom fields, so it was an easy problem to miss. By putting these here it's harder to forget to clear a new `categoryInfo` field in `_registerExtensionPrimitives` before filling it.

kchadha commented on src/engine/runtime.js:
```diff
@@ -830,18 +825,24 @@ class Runtime extends EventEmitter {
             if (extensionInfo.menus.hasOwnProperty(menuName)) {
-                const menuItems = extensionInfo.menus[menuName];
-                const convertedMenu = this._buildMenuForScratchBlocks(menuName, menuItems, categoryInfo);
+                const menuValue = extensionInfo.menus[menuName];
```
I think `menuValue` is a weird name. Maybe we can just go w/`menuInfo` or in cases like this I use 'Desc' as short for descriptor, e.g. `menuDesc` ¯\\_(ツ)_/¯

cwillisf replied:
Ah, yeah... I was going for "name/value" but `menuInfo` is probably better

kchadha commented on src/engine/runtime.js:
```diff
@@ -883,19 +884,21 @@ class Runtime extends EventEmitter {
-        if (typeof menuItems === 'function') {
-            options = menuItems;
+        let menuItems = null;
+        if (typeof menuInfo.items === 'function') {
```
Why would this be a function? O.o

kchadha commented on src/engine/runtime.js:
```diff
@@ -1218,29 +1222,52 @@ class Runtime extends EventEmitter {
+        let shadowType;
+        let fieldName;
+        if (argInfo.menu) {
+            const menuInfo = argInfo.menu && context.categoryInfo.menuInfo[argInfo.menu];
```
We're already in an if for `argInfo.menu` so we don't need the additional check here, right?
```suggestion
            const menuInfo = context.categoryInfo.menuInfo[argInfo.menu];
```

kchadha commented on src/engine/runtime.js:
```diff
@@ -1218,29 +1222,52 @@ class Runtime extends EventEmitter {
+            } else {
+                argJSON.type = 'field_dropdown';
+                argJSON.options = menuInfo.items;
+                valueName = null;
```
why not just initialize these to `null` when declaring them above?

cwillisf replied:
this is a possibly-controversial habit of mine: if I have a variable that gets initialized in one of several ways I generally leave it uninitialized and then make sure to initialize it in each path. That way if something happens to the control flow (like a `break` gets added or something) and results in the possibility of skipping initialization of the variable, an IDE or linter can report that the variable might be uninitialized below. Initializing to `null` makes sure that the variable is definitely initialized but doesn't guarantee that it's initialized to a correct value...

TL;DR: it's a habit of mine that I like for maintenance reasons but if other people don't like it then I'm happy to switch.

kchadha commented on src/engine/runtime.js:
```diff
@@ -1218,29 +1222,52 @@ class Runtime extends EventEmitter {
-            if (fieldType) {
-                context.inputList.push(`<field name="${fieldType}">${defaultValue}</field>`);
-            }
+        // <field> is a text field that the user can type into. Some shadows, like the color picker, don't allow
```
This comment is actually not strictly correct... especially now that we have non-droppable menus...

kchadha commented on src/extension-support/extension-manager.js:
```diff
@@ -310,15 +310,22 @@ class ExtensionManager {
         const menuNames = Object.getOwnPropertyNames(menus);
         for (let i = 0; i < menuNames.length; i++) {
-            const item = menuNames[i];
+            const menuName = menuNames[i];
```
Maybe add a comment here saying we're normalizing the menu definitions to have the structure `{items: [...]}`.

kchadha reviewed (changes requested):
Some minor comments/questions inline.

aeons commented:
Hey, what's the status on this?

kchadha commented:
@aeons, @cwillisf and I are actively working on this as part of a larger project to convert core blocks in scratch into extensions (aka using the extension spec to define them). See #2120 for more info. 

This is a pull request to get these changes into the `e16n` branch, and there will be a subsequent pull request to get the `e16n` branch onto `develop`.

aeons commented:
Alright, thanks.

ErikMejerHansen commented:
@kchadha: Is the plan to finish up all the e16n work and then merge the `e16n`-branch or will there be multiple merges as the work progresses?
This feature would be very useful to have without having to wait for all of the e16n work to finish.

kchadha commented:
@ErikMejerHansen, we will be merging e16n as work progresses. There are currently two PRs out for merging e16n into develop LLK/scratch-gui#4790 and #2161. Our hope is that we will be getting those two PRs merged in soon.

cwillisf committed:
remove redundant check

Co-Authored-By: Karishma Chadha <kchadha@media.mit.edu>

Commit:
respond to code review

cwillisf committed:
Merge branch 'develop' into non-droppable-extension-menus

kchadha committed:
fix extension dropdown menu item conversion
